corehq/apps/hqwebapp/models.py

```diff
@@ -53,3 +52,45 @@ def get_latest_alert(cls):
             return active_alerts[0]
         else:
             return ''
+
+
+class UserAgent(models.Model):
+    MAX_LENGTH = 255
+
+    value = models.CharField(max_length=MAX_LENGTH, db_index=True)
+
+
+class UserAccessLogManager(models.Manager):
+    def create(self, **obj_data):
+        user_agent = obj_data.pop('user_agent', '')
+        if user_agent:
+            user_agent = user_agent[:UserAgent.MAX_LENGTH]
+            agent_ref, _ = UserAgent.objects.get_or_create(value=user_agent)
+            obj_data['user_agent'] = agent_ref
+
+        return super().create(**obj_data)
+
+
+@architect.install('partition', type='range', subtype='date', constraint='month', column='timestamp')
+class UserAccessLog(models.Model):
+    TYPE_LOGIN = 'login'
+    TYPE_LOGOUT = 'logout'
+    TYPE_FAILURE = 'failure'
+
+    ACTIONS = (
+        (TYPE_LOGIN, 'Login'),
+        (TYPE_LOGOUT, 'Logout'),
+        (TYPE_FAILURE, 'Login Failure')
+    )
+
+    user_id = models.CharField(max_length=255, db_index=True)
+    action = models.CharField(max_length=20, choices=ACTIONS)
+    ip = models.GenericIPAddressField(blank=True, null=True)
+    user_agent = models.ForeignKey(UserAgent, null=True, on_delete=models.PROTECT)
+    path = models.CharField(max_length=255, blank=True)
+    timestamp = models.DateTimeField(default=timezone.now)
```

Comment:
Should this use `datetime.utcnow`? I'm not familiar with [`django.utils.timezone.now`](https://docs.djangoproject.com/en/2.2/ref/utils/#django.utils.timezone.now), but it looks like we do not set [`settings.USE_TZ`](https://docs.djangoproject.com/en/2.2/ref/settings/#use-tz), so it will return `datetime.now()`, which is local time.
```suggestion
    timestamp = models.DateTimeField(default=datetime.utcnow)
```
If you accept this suggestion you will also need to add `from datetime import datetime` to this module.

Reply:
I might need to dig into this further. I was following [the docs](https://docs.djangoproject.com/en/3.1/ref/models/fields/#django.db.models.DateField.auto_now_add) again, where it recommends `timezone.now()`. Testing locally with `settings.USE_TZ = False`, both `datetime.now()` and `datetime.utcnow()` look identical